core/trino-spi/src/main/java/io/trino/spi/block/AbstractVariableWidthBlock.java

```diff
@@ -120,6 +120,23 @@ public void writeBytesTo(int position, int offset, int length, BlockBuilder bloc
         blockBuilder.writeBytes(getRawSlice(position), getPositionOffset(position) + offset, length);
     }
 
+    public void writeBytesTo(int[] positions, int positionsCount, BatchByteWriter byteWriter)
+    {
+        int[] lengths = new int[positionsCount];
+        int[] offsets = new int[positionsCount];
+        for (int i = 0; i < positionsCount; i++) {
+            int position = positions[i];
+            if (!isNull(position)) {
+                int length = getSliceLength(position);
+                int offset = getPositionOffset(position);
```

Comment:
We're looking into offset array 3 times here, you can try something like below
```
        int[] offsets = new int[positionsCount];
        int finalLength = 0;
        for (int i = 0; i < positionsCount; i++) {
            int position = positions[i];
            finalLength += getSliceLength(position);
            offsets[i + 1] = finalLength;
        }
```

Reply:
Positions are not necessarily consecutive so the lookups for the given and the subsequent position are necessary.
I expect the compiler to merge two lookups for the current positions into one.
We could check if the positions are consecutive and avoid one lookup at the cost of a branch but I'm not sure it would be beneficial in all cases

Reply:
Okay non-consecutive positions rules that out.